### ISSUE TITLE
Ability to specify guard file destination folder?

### ISSUE THREAD

**Issue by ViSaturn** [rhys-vdw/ts-auto-guard#201]
I want to generate type guards for my source files, but in my tests directory, I checked all your cli options and didn't find something for this. Hope you can help. :)

**Comment by ViSaturn** [comment]
> I think it would be perfectly reasonable to just run find + mv after you've generated the files too. Default behaviour seems fairly sensible to me.
> 
> ```shell
> $ npm run generate-guards && find . -name '*.guard.ts' | xargs -l bash -c 'mv $0 test/guards/folder/'
> ```
I considered your solution before opening this issue, and there's nothing really wrong with it. It just felt weird to create the type guard files in one directory only to immediately move them to another directory rather than just creating them in that directory in the first place.

I think adding a `output-dir` or `out-dir` cli option that recursively loops through the source files & sub-directories, checks for an existing guard file in the output directory and deletes it; later you would output the generated guard.ts files to the output directory rather than in the same directory as the source files would work?

**Comment by ViSaturn** [comment]
> I think it would be perfectly reasonable to just run find + mv after you've generated the files too. Default behaviour seems fairly sensible to me.
> 
> ```shell
> $ npm run generate-guards && find . -name '*.guard.ts' | xargs -l bash -c 'mv $0 test/guards/folder/'
> ```
I also just realized that since the guard files use relative imports to get types that could cause issues if you just were to move the guard file after creation like you suggested.

**Comment by rhys-vdw** [comment]
> I also just realized that since the guard files use relative imports to get types that could cause issues if you just were to move the guard file after creation like you suggested.

Ah true. Although why does it matter if the files are outside of your tests directory? They'll only be included in bundles that import them.

**Comment by download13** [comment]
I'm curious what your use-case is for this. Don't tests usually import types and values from files outside the tests directory?

Generated guards are co-located with their types since that's generally were it makes the most sense to consume them. Is there a reason that the guards need to be in the test directory but their type definitions don't?

Not saying there's no reason for this, but maybe there's a better solution for whatever you're trying to do.

**Comment by ViSaturn** [comment]
> Generated guards are co-located with their types since that's generally were it makes the most sense to consume them. Is there a reason that the guards need to be in the test directory but their type definitions don't?

I use the type definitions in my source code, but I only need to use their type guards in my tests, so I thought it would be cleaner to just have the type guards in my tests directory, but it is still completely possible for me to just use them from the source directory in my tests.

I don't want to use the type guards in my source code because I'm gonna be checking very large interfaces and don't want it to hurt performance, so I thought it'd be fine to only check the interfaces against the type guards during tests.

**Comment by download13** [comment]
> I need the type guards to check very large foreign objects against interfaces so I can know if the API changed (which I do not control) and don't want it to hurt performance

Would it somehow hurt performance to do `import {isType} from '../src/types/type.guard.ts` instead of `import {isType} from './types/type.guard.ts`? That seems like it would be the easiest way to solve your problem. It's also the way idiomatic of testing application code, including type guards.

The guards are side-effect free and tree-shakeable, so any that aren't explicitly used in your main application won't run or even end up in a build bundle. Plus, if you ever do need to use them in the application, it's good to have them accessible.

> I thought it'd be fine to only check the objects against the type guards during tests.

I'd caution against assuming that the properties of an external resource measured at one point in time remain the same later. If your application is running for any significant length of time, the remote API could change out from under you, and it's always possible there will be error output during an exceptional state that does not match the outputs you checked in tests.

That said, if you're okay with possible runtime exceptions, you can use the [shortcircuit option](https://github.com/rhys-vdw/ts-auto-guard#short-circuiting) to use the guards in your application without paying the performance cost of a deep check. Keep in mind this makes the guard always return true so it's basically a glorified cast.

**Comment by rhys-vdw** [comment]
I'm wondering if the real fix here is our discussion from #200, so you could have something like:

```ts
// ./test/guard-types.ts

import { Foo } from "./src/types/Foo"
import { Bar } from "./src/types/Bar"

/* @see {isFoo} ts-auto-guard:type-guard */
export Foo;

/* @see {isBar} ts-auto-guard:type-guard */
export Bar;
```

Then all the guards will appear in `./test/guard-types.guards.ts`.

**Comment by ViSaturn** [comment]
> I'm wondering if the real fix here is our discussion from #200, so you could have something like:
> 
> ```ts
> // ./test/guard-types.ts
> 
> import { Foo } from "./src/types/Foo"
> import { Bar } from "./src/types/Bar"
> 
> /* @see {isFoo} ts-auto-guard:type-guard */
> export Foo;
> 
> /* @see {isBar} ts-auto-guard:type-guard */
> export Bar;
> ```
> 
> Then all the guards will appear in `./test/guard-types.guard.ts`.

This seems like a fairly elegant solution to my issue of only needing type-guards in my tests directory, but my thoughts on this have changed after reading download13's response:

> > I need the type guards to check very large foreign objects against interfaces so I can know if the API changed (which I do not control) and don't want it to hurt performance
> 
> Would it somehow hurt performance to do `import {isType} from '../src/types/type.guard.ts` instead of `import {isType} from './types/type.guard.ts`? That seems like it would be the easiest way to solve your problem. It's also the idiomatic way of testing application code, including type guards.
> 
> The guards are side-effect free and tree-shakeable, so any that aren't explicitly used in your main application won't run or even end up in a build bundle. Plus, if you ever do need to use them in the application, it's good to have them accessible.
> 
> > I thought it'd be fine to only check the objects against the type guards during tests.
> 
> I'd caution against assuming that the properties of an external resource measured at one point in time remain the same later. If your application is running for any significant length of time, the remote API could change out from under you, and it's always possible there will be error output during an exceptional state that does not match the outputs you checked in tests.


I originally thought that having type-guards in the source code wouldn't be beneficial because I am writing an api-wrapper as an npm package, and thought that it wouldn't matter for users to get detailed errors about the type guards and report them to me, when I could just run the tests to find the same errors that they would find and they could have the performance benefit of not having the type guards. But thinking about it now, with the short-circuit option, I could probably just have the type-guards in my source code, let them run normally in my tests, but give the library users the option on whether to short-circuit or not for performance (just a quick reminder again that the library will probably be coming across quite a lot of large foreign json per second depending on their use case for the library).

Thank you both tons for helping me regarding the issue, to me personally the issue has been solved but I'll keep it open until further notice for now.

**Comment by rhys-vdw** [comment]
> Thank you both tons for helping me regarding the issue, to me personally the issue has been solved but I'll keep it open until further notice for now.

Pleasure. Glad you found your solution! I think I'll close this one for now then, since the fix to #200 should offer neat solution for anyone facing the same issue.

Good luck with your library.
